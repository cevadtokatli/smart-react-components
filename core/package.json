{
  "name": "@smart-react-components/core",
  "version": "2.0.0-beta-0",
  "license": "MIT",
  "main": "index.js",
  "description": "SRC core includes the theme, Styled properties, and Element components.",
  "keywords": [
    "javascript",
    "typescript",
    "react",
    "styled-components"
  ],
  "author": {
    "name": "Cevad Tokatlı",
    "email": "cevadtokatli@hotmail.com"
  },
  "contributors": [
    {
      "name": "Cemil Tokatlı",
      "email": "cemil@tokatli.me"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cevadtokatli/smart-react-components/tree/v2/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "yarn clear && tsc --outDir lib && yarn copy",
    "clear": "rm -rf lib",
    "copy": "cp package.json lib/package.json && cp README.md lib/README.md && cp ../LICENSE lib/LICENSE",
    "dev": "tsc -w --outDir lib",
    "lint": "eslint src",
    "release": "../scripts/release.sh core",
    "test": "jest"
  },
  "devDependencies": {
    "@types/color": "3.0.3",
    "@types/enzyme": "3.10.12",
    "@types/jest": "29.0.1",
    "@types/react": "18.0.19",
    "@types/react-dom": "18.0.6",
    "@types/styled-components": "5.1.26",
    "@wojtekmaj/enzyme-adapter-react-17": "0.6.7",
    "enzyme": "3.11.0",
    "eslint": "8.0.1",
    "jest": "29.0.3",
    "jsdom": "20.0.0",
    "jsdom-global": "3.0.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "styled-components": "5.3.5",
    "ts-jest": "29.0.0",
    "typescript": "4.8.3"
  },
  "dependencies": {
    "color": "4.2.3"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "styled-components": "^5.3.5"
  }
}
