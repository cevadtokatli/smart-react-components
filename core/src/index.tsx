export * from './theme'
export { default as createTheme } from './theme'
export * from './styled-props'
export * from './Element'
