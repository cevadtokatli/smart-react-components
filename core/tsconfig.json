{
  "extends": "../tsconfig.json",
  "include": ["src/**/*.tsx"],
  "exclude": ["node_modules", "src/**/*.test.tsx"]
}
