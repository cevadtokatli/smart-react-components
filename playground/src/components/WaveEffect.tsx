import Div from '@smart-react-components/core/Element/Div'
import WaveEffect from '@smart-react-components/ui/WaveEffect'
import React from 'react'
import styled from 'styled-components'

const Button = styled(Div).attrs({
  borderRadius: '.125rem',
  cursor: 'default',
  display: 'inline-block',
  marginBottom: '$length.3',
  marginRight: '$length.3',
  padding: '$length.2 $length.3', 
})``

export const Palette = () => (
  <Div display="flex" flexWrap="wrap">
    <WaveEffect><Button background="!dark" color="!light">Light</Button></WaveEffect>
    <WaveEffect palette="dark"><Button>Dark</Button></WaveEffect>
    <WaveEffect palette="primary"><Button>Primary</Button></WaveEffect>
    <WaveEffect palette="secondary"><Button>Secondary</Button></WaveEffect>
    <WaveEffect palette="success"><Button>Success</Button></WaveEffect>
    <WaveEffect palette="danger"><Button>Danger</Button></WaveEffect>
    <WaveEffect palette="warning"><Button>Warning</Button></WaveEffect>
    <WaveEffect palette="info"><Button>Info</Button></WaveEffect>
  </Div>
)

export const Duration = () => <WaveEffect duration={1000}><Button background="!dark" color="!light">Light</Button></WaveEffect>

