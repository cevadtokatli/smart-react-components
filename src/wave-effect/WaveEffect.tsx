/// <reference path="../modules/global.d.tsx" />
import React from "react"
import {findDOMNode} from "react-dom"
import {createGlobalStyle} from "styled-components"
import DOMHelper from "../helper/DOMHelper"
import DV from "../default-value"

let id = 0

const getID = () => process.env.NODE_ENV !== "test" ? id++ : 0

if(typeof global !== "undefined")
    global.srcWaveEffectCleanID = () => id = 0

interface GlobalStyleProps {
    id: number
    type: string
}

const GlobalStyle = createGlobalStyle<GlobalStyleProps>(({theme,id,type}) => `
    *[data-src-wave-effect="${id}"] {
        position: relative;
        overflow: hidden;
        -webkit-tap-highlight-color: transparent;

        .src-wave-ripple {
            position: absolute;
            width: 100px;
            height: 100px;
            margin-top: -50px;
            margin-left: -50px;
            border-radius: 50%;
            background: ${(theme as any).src.type[type].waveEffect.color};
            pointer-events: none;
            transform: scale(0);
            opacity: 0;
            transition: ${(theme as any).src.waveEffect.transition};
            transition-property: transform, opacity;
        }
    }
`)

export interface Props {
    children: JSX.Element
    type?: string
}

export default class WaveEffect extends React.Component<Props> {
    public static defaultProps = {
        type: "light"
    }

    private active: boolean = false
    private lastRippleEl: HTMLElement = null
    private id: number = getID()

    componentDidMount() {
        const el = findDOMNode(this) as HTMLElement
        if(el)
            el.setAttribute("data-src-wave-effect", this.id.toString())
    }

    addEventListener() {
        DOMHelper.addEventListener(window, ["mousemove"], this.mouseMove)
        DOMHelper.addEventListener(window, ["touchmove"], this.touchMove)
        DOMHelper.addEventListener(window, ["mouseup"], this.mouseUp)
        DOMHelper.addEventListener(window, ["touchend"], this.touchEnd)
    }

    removeEventListener() {
        DOMHelper.removeEventListener(window, ["mousemove"], this.mouseMove)
        DOMHelper.removeEventListener(window, ["touchmove"], this.touchMove)
        DOMHelper.removeEventListener(window, ["mouseup"], this.mouseUp)
        DOMHelper.removeEventListener(window, ["touchend"], this.touchEnd)
    }

    removeRippleElement = (rippleEl:HTMLElement): void => {
        const ignored = rippleEl.offsetWidth
        rippleEl.style.opacity = "0"

        setTimeout(() => {
            const containerEl = findDOMNode(this) as HTMLElement
            containerEl.removeChild(rippleEl)

            if(this.lastRippleEl == rippleEl)
                this.lastRippleEl = null
        }, DV.WAVE_EFFECT_TRANSITION_DURATION)
    }

    mouseDown = (e:MouseEvent&TouchEvent): void => {
        if(DOMHelper.isMobile)
            return
        if(this.props.children.props.onMouseDown)
            this.props.children.props.onMouseDown(e)
        this.start(e)
    }

    touchStart = (e:MouseEvent&TouchEvent): void => {
        if(this.props.children.props.onTouchStart)
            this.props.children.props.onTouchStart(e)
        this.start(e)
    }

    start = (e:MouseEvent&TouchEvent): void => {
        if(e.button == 2)
            return

        if(!this.active) {
            this.active = true
            this.addEventListener()
        }

        const containerEl = findDOMNode(this) as HTMLElement
        const rippleEl = this.lastRippleEl = document.createElement("div")
        const rect = containerEl.getBoundingClientRect()
        const scale = (containerEl.clientWidth / 100) * 3

        let left
        let top
        if(DOMHelper.isMobile) {
            const t = e.touches[0]
            left = (t.pageX - window.pageXOffset) - rect.left
            top =  (t.pageY - window.pageYOffset) - rect.top
        } else {
            left = e.clientX - rect.left
            top = e.clientY - rect.top
        }

        rippleEl.classList.add("src-wave-ripple")
        rippleEl.setAttribute("data-src-end-time", (new Date().getTime() + 275).toString())
        containerEl.appendChild(rippleEl)
        const ignored = rippleEl.offsetWidth

        rippleEl.setAttribute("style", `
            left: ${left}px;
            top: ${top}px;
            transform: scale(${scale});
            opacity: 1;
        `)

        setTimeout(() => {
            if(!this.active || this.lastRippleEl != rippleEl)
                this.removeRippleElement(rippleEl)
        }, 275)
    }

    mouseMove = (e:MouseEvent): void => {
        if(DOMHelper.isMobile)
            return
        if(this.props.children.props.onMouseMove)
            this.props.children.props.onMouseMove(e)
        this.move(e)
    }

    touchMove = (e:MouseEvent): void => {
        if(this.props.children.props.onTouchMove)
            this.props.children.props.onTouchMove(e)
        this.move(e)
    }

    move = (e:MouseEvent): void => {
        if(this.active && this.lastRippleEl) {
            const target = e.target as HTMLElement
            const containerEl = findDOMNode(this) as HTMLElement
            if(target != containerEl && !containerEl.contains(target)) {
                this.active = false
                this.removeEventListener()
                this.removeRippleElement(this.lastRippleEl)
            }
        }
    }

    mouseUp = (e:MouseEvent): void => {
        if(DOMHelper.isMobile)
            return
        if(this.props.children.props.onMouseUp)
            this.props.children.props.onMouseUp(e)
        this.end()
    }

    touchEnd = (e:MouseEvent): void => {
        if(this.props.children.props.onTouchEnd)
            this.props.children.props.onTouchEnd(e)
        this.end()
    }

    end = (): void => {
        if(this.active) {
            this.active = false
            this.removeEventListener()
            
            if(this.lastRippleEl) {
                const date = parseInt(this.lastRippleEl.getAttribute("data-src-end-time"))

                if(date < new Date().getTime())
                    this.removeRippleElement(this.lastRippleEl)
            }
        }
    }

    render() {
        return (
            <>
                <GlobalStyle id={this.id} type={this.props.type} />
                {React.cloneElement(this.props.children, {
                    "data-src-wave-effect": this.id.toString(),
                    onMouseDown: this.mouseDown,
                    onTouchStart: this.touchStart
                })}
            </>
        )
    }
}