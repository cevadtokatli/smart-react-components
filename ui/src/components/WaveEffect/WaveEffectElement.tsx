import Div from '@smart-react-components/core/Element/Div'
import { PaletteProp } from '@smart-react-components/core/types'
import styled from 'styled-components'

interface Props {
  duration: number
  left: number
  palette: PaletteProp
  scale: number
  top: number
}

export default styled(Div)<Props>(({ duration, left, palette, scale, theme, top }) => `
  background: ${theme.$.palette[palette].waveEffect};
  border-radius: 50%;
  height: 100px;
  left: ${left}px;
  margin: -50px 0 0 -50px;
  opacity: .75;
  pointer-events: none;
  position: absolute;
  transform: scale(${scale});
  top: ${top}px;
  width: 100px;

  &.src-wave-effect-show {
    opacity: 0;
    transform: scale(0);
  }

  &.src-wave-effect-show-active {
    opacity: .75;
    transform: scale(${scale});  
  }

  &.src-wave-effect-hide-active {
    opacity: 0;
  }

  &.src-wave-effect-show-active,
  &.src-wave-effect-hide-active {
    transition: ease-out ${duration}ms;
    transition-property: opacity, transform;
  }
`)
