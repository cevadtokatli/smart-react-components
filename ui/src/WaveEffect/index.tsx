import { Theme } from '@smart-react-components/core'
import { PaletteProp } from '@smart-react-components/core/types'
import { addEventListener, isMobile, removeEventListener } from '@smart-react-components/core/util/dom'
import CSSTransitionGroup from '@smart-react-components/transition/CSSTransitionGroup'
import React from 'react'
import { ThemeContext } from 'styled-components'
import WaveEffectElement from './WaveEffectElement'

interface EffectItem {
  element: JSX.Element
  endTime: number
}

export interface Props {
  children: JSX.Element
  duration?: number
  palette?: PaletteProp
}

const WaveEffect = React.forwardRef<HTMLDivElement, Props>(({ children, duration, palette = 'light' }, forwardRef) => {
  const theme: Theme = React.useContext(ThemeContext)
  const [effectItems, setEffectItems] = React.useState<EffectItem[]>(() => [])
  const isProcessing = React.useRef(false)
  const ref = (forwardRef ?? (children as any).ref) as React.MutableRefObject<HTMLElement> ?? React.useRef<HTMLElement>(null)

  const getDuration = (): number => duration ?? theme.$.transition.waveEffectDuration

  React.useEffect(() => {
    if (effectItems.length > 0) {
      addEventListener(window, [isMobile ? 'touchmove' : 'mousemove'], handleMove)
      addEventListener(window, [isMobile ? 'touchend' : 'mouseup'], handleEnd)

      return () => {
        removeEventListener(window, [isMobile ? 'touchmove' : 'mousemove'], handleMove)
        removeEventListener(window, [isMobile ? 'touchend' : 'mouseup'], handleEnd)
      }
    }
  }, [effectItems])

  const handleMouseDown = (e: MouseEvent) => {
    children.props?.onMouseDown?.(e)

    if (!isMobile && e.button === 0) {
      start(e)
    }
  }

  const handleTouchStart = (e: TouchEvent) => {
    children.props?.onTouchStart?.(e)

    start(e)
  }

  const start = (e: MouseEvent | TouchEvent) => {
    isProcessing.current = true
    const rect = ref.current.getBoundingClientRect()
    const scale = (ref.current.clientWidth / 100) * 2
    let left: number
    let top: number

    if (isMobile) {
      const t = (e as TouchEvent).touches[0]
      left = (t.pageX - window.pageXOffset) - rect.left
      top = (t.pageY - window.pageYOffset) - rect.top
    } else {
      left = (e as MouseEvent).clientX - rect.left
      top = (e as MouseEvent).clientY - rect.top
    }

    setEffectItems([
      ...effectItems,
      {
        element: (
          <WaveEffectElement
            key={new Date().getTime()}
            duration={getDuration() / 2}
            left={left}
            palette={palette}
            scale={scale}
            top={top}
          />
        ),
        endTime: new Date().getTime() + getDuration(),
      },
    ])
  }

  const handleMove = (e: MouseEvent | TouchEvent) => {
    if (e.target !== ref.current && !ref.current.contains(e.target as Node)) {
      handleEnd()
    }
  }

  const handleEnd = () => {
    isProcessing.current = false
    const time = new Date().getTime()
    setEffectItems(effectItems.filter(e => e.endTime >= time))
  }

  const handleAfterShow = (child: JSX.Element) => {
    if (
      !isProcessing.current
      || (
        isProcessing.current
        && effectItems.findIndex(e => e.element.key === child.key) + 1 < effectItems.length
      )
    ) {
      setEffectItems(effectItems.filter(e => e.element.key !== child.key))
    }
  }

  return React.cloneElement(children, {
    children: (
      <>
        { children.props.children }
        <CSSTransitionGroup
          afterShow={handleAfterShow}
          className="src-wave-effect"
          duration={getDuration() / 2}
        >
          { effectItems.map(e => e.element) }
        </CSSTransitionGroup>
      </>
    ),
    onMouseDown: handleMouseDown,
    onTouchStart: handleTouchStart,
    overflow: 'hidden',
    position: 'relative',
    ref,
    webkitTapHighlightColor: 'transparent',
  })
})

export default WaveEffect
