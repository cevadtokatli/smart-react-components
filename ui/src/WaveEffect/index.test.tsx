import { wrapTheme } from '@smart-react-components/core/test'
import { act, createEvent, fireEvent, render, waitFor } from '@testing-library/react' 
import React from 'react'
import WaveEffect from './'

describe('<WaveEffect />', () => {
  let Component

  beforeAll(() => {
    Component = () => <WaveEffect><div>Wave Effect</div></WaveEffect>
  })

  it('should render wave effect', async () => {
    const { container, asFragment, rerender } = render(wrapTheme(<Component />))
    const node = container.querySelector('div')!

    await waitFor(() => {
      fireEvent(node, createEvent.mouseDown(node))
    })

    rerender(wrapTheme(<Component />))

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not render wave effect', () => {
    const { asFragment } = render(wrapTheme(<Component />))
    expect(asFragment()).toMatchSnapshot()
  })
})
