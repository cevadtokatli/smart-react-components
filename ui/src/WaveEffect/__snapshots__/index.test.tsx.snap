// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WaveEffect /> should not render wave effect 1`] = `
<DocumentFragment>
  <div
    overflow="hidden"
    position="relative"
    webkittaphighlightcolor="transparent"
  >
    Wave Effect
  </div>
</DocumentFragment>
`;

exports[`<WaveEffect /> should render wave effect 1`] = `
<DocumentFragment>
  .c0 {
  left: 0;
  top: 0;
}

.c1 {
  background: radial-gradient(rgba(241,245,247,0.2) 0,rgba(241,245,247,0.3) 40%,rgba(241,245,247,0.4) 50%,rgba(241,245,247,0.5) 60%,rgba(241,245,247,0) 70%);
  border-radius: 50%;
  height: 100px;
  left: 0px;
  margin: -50px 0 0 -50px;
  opacity: .75;
  pointer-events: none;
  position: absolute;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  top: 0px;
  width: 100px;
}

.c1.src-wave-effect-show {
  opacity: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c1.src-wave-effect-show-active {
  opacity: .75;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c1.src-wave-effect-hide-active {
  opacity: 0;
}

.c1.src-wave-effect-show-active,
.c1.src-wave-effect-hide-active {
  -webkit-transition: ease-out 375ms;
  transition: ease-out 375ms;
  -webkit-transition-property: opacity,-webkit-transform;
  -webkit-transition-property: opacity,transform;
  transition-property: opacity,transform;
}

<div
    overflow="hidden"
    position="relative"
    webkittaphighlightcolor="transparent"
  >
    Wave Effect
    <div
      class="c0 c1"
      scale="0"
    />
  </div>
</DocumentFragment>
`;
